### PR TITLE
Add statistics collection

### PR DESCRIPTION
**Pull request by kononovk**
Add statistics collection in thread_data class

Co-Authored-By: Zakhar Zakharov <zakhar.zakharov@intel.com>

### CODE REVIEW

File: src/tbb/arena_slot.cpp [hunk 1]
```diff
@@ -59,6 +59,9 @@ d1::task* arena_slot::get_task_impl(size_t T, execution_data_ext& ed, bool& task
 }
 
 d1::task* arena_slot::get_task(execution_data_ext& ed, isolation_type isolation) {
+#ifdef __TBB_STATISTICS
+    governor::get_thread_data()->add_statistics(thread_data::Statistics::GET_TASK);
+#endif /* __TBB_STATISTICS */
```
Comment (alexey-katranov) [inline src/tbb/arena_slot.cpp:64]: What about to wrap this three lines into one, e.g. marco or function (that is empty if __TBB_STATISTICS is not set)?

File: src/tbb/arena_slot.cpp [hunk 2]
```diff
@@ -142,12 +145,23 @@ d1::task* arena_slot::get_task(execution_data_ext& ed, isolation_type isolation)
...
 
 d1::task* arena_slot::steal_task(arena& a, isolation_type isolation) {
+#ifdef __TBB_STATISTICS
+    governor::get_thread_data()->add_statistics(thread_data::Statistics::TRY_STEAL_TASK);
+#endif /* __TBB_STATISTICS */
     d1::task** victim_pool = lock_task_pool();
     if (!victim_pool) {
+#ifdef __TBB_STATISTICS
+        governor::get_thread_data()->add_statistics(thread_data::Statistics::VICTIM_POOL_EMPTY);
+#endif /* __TBB_STATISTICS */
```
Comment (alexey-katranov) [inline src/tbb/arena_slot.cpp:164]: Taking TLS (`governor::get_thread_data()`) is not cheapest operation, in many places `thread_data` is available (or consider passing it trough the function arguments).  @pavelkumbrasev , what do you think?

File: src/tbb/thread_data.h [hunk 3]
```diff
@@ -204,6 +213,74 @@ class thread_data : public ::rml::job
...
+    using ArrayStat = std::array<std::uint32_t, static_cast<std::size_t>(thread_data::Statistics::SIZE)>;
+    ArrayStat statistics {}; // main thread statistics object
+
+    class CollectStatistics : no_copy {
+    public:
+        CollectStatistics() = default;
+
+        CollectStatistics(CollectStatistics&&) = delete;
+
+        CollectStatistics& operator=(CollectStatistics&&) = delete;
```
Comment (alexey-katranov) [inline src/tbb/thread_data.h:241]: no_copy and ` CollectStatistics(CollectStatistics&&) = delete;`  seems as duplication. Moreover, `std::mutex` is not copiable.

File: src/tbb/thread_data.h [hunk 4]
```diff
@@ -204,6 +213,74 @@ class thread_data : public ::rml::job
...
+    };
+
+    void add_statistics(Statistics stat_type) {
+        ++statistics[static_cast<std::size_t>(stat_type)];
+    }
+private:
+    using ArrayStat = std::array<std::uint32_t, static_cast<std::size_t>(thread_data::Statistics::SIZE)>;
+    ArrayStat statistics {}; // main thread statistics object
+
+    class CollectStatistics : no_copy {
```
Comment (alexey-katranov) [inline src/tbb/thread_data.h:235]: Be consistent with code style (snake_style is preferable)

File: src/tbb/task_dispatcher.h [hunk 5]
```diff
@@ -161,6 +161,9 @@ inline d1::task* task_dispatcher::steal_or_get_critical(
     if (d1::task* t = a.steal_task(arena_index, random, ed, isolation)) {
         ed.context = task_accessor::context(*t);
         ed.isolation = task_accessor::isolation(*t);
+#ifdef __TBB_STATISTICS
+        governor::get_thread_data()->add_statistics(thread_data::Statistics::STEAL_TASK);
+#endif /* __TBB_STATISTICS */
```
Comment (ivankochin) [inline src/tbb/task_dispatcher.h]: I would prefer to move successful stealing statistic calculation inside the `arena_slot::steal_task` to keep all stealing-related statistics in one place (if it is possible).

Also, don't we want to calculate only the tasks that weren't respawned after finding a critical task? @pavelkumbrasev @alexey-katranov what do you think about it?

Comment (kononovk) [reply src/tbb/task_dispatcher.h]: Hello, check new version of add_statistics, please

**Pull request closed without merging**